src/hotspot/share/runtime/memprofiler.cpp

```diff
@@ -115,6 +115,9 @@ void MemProfiler::do_trace() {
       resource_memory_usage += cur->resource_area()->size_in_bytes();
     }
 
+    // used() calls G1Allocator::used_in_alloc_regions() when G1 enabled,
+    // it checks whether Heap_lock was owned on this thread's behalf.
+    G1GC_ONLY(MutexLocker ml(Heap_lock);)
```

Comment:
This seems really out of place - this code should not need to know anything about the specific locking requirements of different GC's. That should be handled inside the appropriate chunk of GC code.